File: frontend/src/app/core/services/community-goal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CommunityGoal {
  id: number;
  title: string;
  targetKm: number;
  progressKm: number;
  progressPercent: number;
  startDate: string;
  endDate: string;
  active: boolean;
  completed: boolean;
}

export interface CreateGoalRequest {
  title: string;
  targetKm: number;
  startDate: string;
  endDate: string;
}

@Injectable({ providedIn: 'root' })
export class CommunityGoalService {
  constructor(private http: HttpClient) {}

  getGoal(communityId: number): Observable<CommunityGoal> {
    return this.http.get<CommunityGoal>(`/api/communities/${communityId}/goal`);
  }

  setGoal(communityId: number, req: CreateGoalRequest): Observable<CommunityGoal> {
    return this.http.post<CommunityGoal>(`/api/communities/${communityId}/goal`, req);
  }
}

File: backend/src/main/java/com/runhub/communities/service/CommunityGoalService.java
package com.runhub.communities.service;

import com.runhub.communities.dto.CommunityGoalDto;
import com.runhub.communities.dto.CreateGoalRequest;
import com.runhub.communities.model.CommunityGoal;
import com.runhub.communities.repository.CommunityGoalRepository;
import com.runhub.communities.repository.CommunityMemberRepository;
import com.runhub.running.repository.ActivityRepository;
import com.runhub.users.model.User;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class CommunityGoalService {

    private final CommunityGoalRepository goalRepository;
    private final CommunityMemberRepository memberRepository;
    private final ActivityRepository activityRepository;

    public Optional<CommunityGoalDto> getCurrentGoal(Long communityId) {
        return goalRepository.findTopByCommunityIdOrderByCreatedAtDesc(communityId)
                .map(g -> toDto(g, communityId));
    }

    public CommunityGoalDto createGoal(Long communityId, CreateGoalRequest req, User admin) {
        CommunityGoal goal = new CommunityGoal();
        goal.setCommunityId(communityId);
        goal.setTitle(req.getTitle());
        goal.setTargetKm(req.getTargetKm());
        goal.setStartDate(req.getStartDate() != null ? req.getStartDate() : LocalDate.now());
        goal.setEndDate(req.getEndDate() != null ? req.getEndDate() : LocalDate.now().plusDays(30));
        goal.setCreatedBy(admin.getId());
        return toDto(goalRepository.save(goal), communityId);
    }

    private CommunityGoalDto toDto(CommunityGoal g, Long communityId) {
        List<Long> memberIds = memberRepository.findByCommunityId(communityId)
                .stream().map(m -> m.getUser().getId()).toList();

        double progress = 0;
        if (!memberIds.isEmpty()) {
            Double sum = activityRepository.sumDistanceByUserIdsAndDateRange(memberIds, g.getStartDate(), g.getEndDate());
            progress = sum != null ? sum : 0;
        }

        double percent = g.getTargetKm() > 0 ? Math.min(100.0, progress / g.getTargetKm() * 100) : 0;
        LocalDate today = LocalDate.now();
        boolean active = !today.isBefore(g.getStartDate()) && !today.isAfter(g.getEndDate());

        return CommunityGoalDto.builder()
                .id(g.getId())
                .title(g.getTitle())
                .targetKm(g.getTargetKm())
                .progressKm(Math.round(progress * 10.0) / 10.0)
                .progressPercent(Math.round(percent * 10.0) / 10.0)
                .startDate(g.getStartDate())
                .endDate(g.getEndDate())
                .active(active)
                .completed(progress >= g.getTargetKm())
                .build();
    }
}
